lib/clustering/mean_shift.cpp

```diff
@@ -6,39 +6,113 @@
+	bool has_converged = false;
+  int iterations = 0; //for logging
+
+	// The MeanShift loop
+	while (!has_converged) {
+		iterations++;
+		if (iterations > MAX_ITERATIONS) {
+			BOOST_LOG_TRIVIAL(error) << "MeanShift: max number of iterations exceeded without convergence";
+			break;
+		}
+		// perform 1 iteration on each mode
+		for (int i=0; i<prev.size(); i++) {
+			curr[i] = iterate_mode(prev[i], prev);
+		}
+
+		// Merge step
+		// For every mode..
+		for (int i=0; i<curr.size(); i++) {
+			// Check modes we haven't already executed the (i)-loop for
+			for (int j=i+1; j<curr.size(); j++) {
+				Eigen::VectorXd diff = curr[i] - curr[j];
+
+				if (diff.norm() < MERGE_THRESHOLD) {
+					// Merge clusters. Erase one of the modes corresponding to the clusters and append points belonging to j to cluser of i
+					for (int k=0; k < clusters[j].points().size(); k++) {
+						clusters[i].points().push_back(clusters[j].points()[k]);
+					}
+					//BOOST_LOG_TRIVIAL(debug) << "Merging cluster " << j << " into cluster " << i << ", distance " << diff.norm();
+					clusters.erase(clusters.begin() + j);
+					curr.erase(curr.begin() + j);
 
-	std::vector<Cluster> clustering = std::vector<Cluster>(_k);
+					break;
+				}
+			}
+		}
 
-  //Initialize RNG
-	srand(time(NULL));
+		if (clusters.size() != curr.size())
+			BOOST_LOG_TRIVIAL(warning) << "Cluster count does not equal mode count. (" << clusters.size() << " != " << curr.size() << ")";
 
-	//Assign random cluster to each point in cloud
-	for(PointIndex i=0; i<cloud.size(); i++) {
-		//Assign index i to random cluster in clustering
-		int cluster = rand() % _k;
-		clustering[cluster].points().push_back(i);
+		// Check convergence
+		if (prev.size() == curr.size()) {
+			double total_movement = 0;
+			for (int i=0; i < curr.size(); i++) {
+				total_movement += (curr[i]-prev[i]).norm();
+			}
+			if (total_movement <= CONVERGENCE_THRESHOLD) has_converged = true;
+		}
+		prev = curr;
 	}
+	BOOST_LOG_TRIVIAL(trace) << "MeanShift converged! #Clusters: " << curr.size() << " #Iterations: " << iterations;
 
-	//get rid of empty clusters
-	int nextInsert = 0;
-  for(int i = 0; i < clustering.size(); i++){
-    if(!clustering[i].points().empty()){
-      clustering[nextInsert] = clustering[i];
-      nextInsert++;
-    }
-  }
-	clustering.erase(clustering.begin() + nextInsert, clustering.end());
-
-	return clustering;
+
+	return clusters;
+}
+
+double MeanShift::apply_gaussian_kernel(const Eigen::VectorXd point, const Eigen::VectorXd mean) const {
+
+	// check if dimensions match
+	if (mean.size() != point.size()) BOOST_LOG_TRIVIAL(error) << "Vector dimensions don't match";
+
+	// Euclidean distance squared btw. mean of kernel and point of interest
+	const double d = (mean-point).squaredNorm();
+
+	// Apply gaussian distribution
+	return exp(-d/(2*_window_size));
+}
+
+Eigen::VectorXd MeanShift::iterate_mode(const Eigen::VectorXd mode, const std::vector<Eigen::VectorXd>& state) const {
```

Comment:
`state` seems a bit un-sharp to me, maybe `referencePoints` or `fixedPoints`?

Reply:
Hmm, I agree that the naming is not optimal. But they are not really points in the sense of PointCloud-points, they are just the other modes (cluster centers). Maybe other_modes? ... Haven't found a name that I'm happy with yet.

Reply:
postponing to skype...